src/number_theory/cyclotomic/basic.lean

```diff
@@ -0,0 +1,209 @@
+given by roots of unity of order in `S`. -/
+lemma union_left [h : is_cyclotomic_extension T A B] (hS : S ⊆ T) :
+  is_cyclotomic_extension S A (adjoin A { b : B | ∃ a : ℕ+, a ∈ S ∧ b ^ (a : ℕ) = 1 }) :=
+begin
+  refine ⟨λ n hn, _, λ b, _⟩,
+  { obtain ⟨b, hb⟩ := ((is_cyclotomic_extension_iff _ _ _).1 h).1 (hS hn),
+    refine ⟨⟨b, subset_adjoin ⟨n, hn, _⟩⟩, _⟩,
+    { rw [aeval_def, eval₂_eq_eval_map, map_cyclotomic, ← is_root.def] at hb,
+      suffices : (X ^ (n : ℕ) - 1).is_root b,
+      { simpa [sub_eq_zero] using this },
+      exact hb.dvd (cyclotomic.dvd_X_pow_sub_one _ _) },
+      rwa [← subalgebra.coe_eq_zero, aeval_subalgebra_coe, subtype.coe_mk] },
+  { convert mem_top,
+    rw ← adjoin_adjoin_coe_preimage,
+    simp,
+    norm_cast, }
+end
+
+end basic
+
+section fintype
+
+lemma finite_of_singleton [is_domain B] [h : is_cyclotomic_extension {n} A B] : finite A B :=
+begin
+  classical,
+  rw [module.finite_def, ← top_to_submodule, ← ((iff_adjoin_eq_top _ _ _).1 h).2],
+  refine fg_adjoin_of_finite _ (λ b hb, _),
+  { simp only [mem_singleton_iff, exists_eq_left],
+    have : {b : B | b ^ (n : ℕ) = 1} = (nth_roots n (1 : B)).to_finset :=
+      set.ext (λ x, ⟨λ h, by simpa using h, λ h, by simpa using h⟩),
+    rw [this],
+    exact (nth_roots ↑n 1).to_finset.finite_to_set },
+  { simp only [mem_singleton_iff, exists_eq_left, mem_set_of_eq] at hb,
+    refine ⟨X ^ (n : ℕ) - 1, ⟨monic_X_pow_sub_C _ n.pos.ne.symm, by simp [hb]⟩⟩ }
+end
+
+/-- If `S` is finite and `is_cyclotomic_extension S A B`, then `B` is a finite `A`-algebra. -/
+lemma finite [is_domain B] [h₁ : fintype S] [h₂ : is_cyclotomic_extension S A B] : finite A B :=
+begin
+  unfreezingI {revert h₂ A B},
+  refine set.finite.induction_on (set.finite.intro h₁) (λ A B, _) (λ n S hn hS H A B, _),
+  { introsI _ _ _ _ _,
+    refine module.finite_def.2 ⟨({1} : finset B), _⟩,
+    simp [← top_to_submodule, ← empty, to_submodule_bot] },
+  { introsI _ _ _ _ h,
+    haveI : is_cyclotomic_extension S A (adjoin A { b : B | ∃ (n : ℕ+),
+      n ∈ S ∧ b ^ (n : ℕ) = 1 }) := union_left _ (insert n S) _ _ (subset_insert n S),
+    haveI := H A (adjoin A { b : B | ∃ (n : ℕ+), n ∈ S ∧ b ^ (n : ℕ) = 1 }),
+    haveI : finite (adjoin A { b : B | ∃ (n : ℕ+), n ∈ S ∧ b ^ (n : ℕ) = 1 }) B,
+    { rw [← union_singleton] at h,
+      letI := @union_right S {n} A B _ _ _ h,
+      exact finite_of_singleton n _ _ },
+    exact finite.trans (adjoin A { b : B | ∃ (n : ℕ+), n ∈ S ∧ b ^ (n : ℕ) = 1 }) _ }
+end
+
+/-- A cyclotomic finite extension of a number field is a number field. -/
+lemma number_field [h : number_field K] [fintype S] [is_cyclotomic_extension S K L] :
+  number_field L :=
+{ to_char_zero := char_zero_of_injective_algebra_map (algebra_map K L).injective,
+  to_finite_dimensional := @finite.trans _ K L _ _ _ _
+    (@rat.algebra_rat L _ (char_zero_of_injective_algebra_map (algebra_map K L).injective)) _ _
```

Comment:
If you merge against master I think this will be needed
```suggestion
    (@algebra_rat L _ (char_zero_of_injective_algebra_map (algebra_map K L).injective)) _ _
```